### PR TITLE
Fix #298: Allow users to specify custom world types in server.yml; Allow registration of custom chunk generators

### LINKED ISSUE #298: Add the ability to specify world type in server.yml
## Expected Behavior
I should be able to specify the world type *(normal, flat, etc)* in the server.yml.  I currently cannot do this as the option does not exist. This makes it harder for people to start building stuff which requires a flat world, as they would need to code a plugin to generate one for them.

### PR DESCRIPTION
**Pull request by ghost**
This pull request completely resolves issue #298 .

Features implemented:
- [X] `server.yml` now ships with `chunkGenerator` property in each world configuration
- [X] Users are now able to choose between standard shipped world types `normal`, `flat` and `void`
- [X] Plugin developers can now register custom `ChunkGenerator` implementations by calling `ChunkGenerator.getRegistry().registerGenerator(String, Class<? extends ChunkGenerator>)`
- [X] Plugin developers can override already registered chunk generators like the chunk generator for `normal`

When merging please squash all commits

### CODE REVIEW

File: gomint-api/src/main/java/io/gomint/world/generator/ChunkGenerator.java [hunk 1]
```diff
@@ -1,24 +1,43 @@
...
 import lombok.Getter;
 
 /**
+ * @author Clockw1seLrd
  * @author geNAZt
  * @version 1.0
  */
 public abstract class ChunkGenerator {
 
+    private static ChunkGeneratorRegistry registry;
```
Comment (geNAZt) [inline gomint-api/src/main/java/io/gomint/world/generator/ChunkGenerator.java]: I don't like the fact that this is static. Maybe this should be integrated in Gomint.instance().getChunkGeneratorRegistry() ?

File: gomint-api/src/main/java/io/gomint/world/generator/ChunkGeneratorRegistry.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import java.util.Collection;
+import java.util.Optional;
+
+/**
+ * @author Clockw1seLrd
+ * @version 1.0
+ */
+public interface ChunkGeneratorRegistry {
+
+    ChunkGenerator createGenerator( String name, World world, GeneratorContext ctx );
```
Comment (geNAZt) [inline gomint-api/src/main/java/io/gomint/world/generator/ChunkGeneratorRegistry.java]: Why would a plugin want to create a generator without being able to use the chunks generated properly?

File: gomint-api/src/main/java/io/gomint/world/generator/ChunkGeneratorRegistry.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+    boolean registerGenerator( String name, Class<? extends ChunkGenerator> generatorClass );
+
+    boolean isGeneratorAvailable( String name );
+
+    Optional<Class<? extends ChunkGenerator>> getGeneratorClass( String name );
```
Comment (geNAZt) [inline gomint-api/src/main/java/io/gomint/world/generator/ChunkGeneratorRegistry.java]: Why use a optional? All other API methods require the developer to check for non null. This should be consistent for all API calls

File: gomint-server/src/main/java/io/gomint/server/world/WorldAdapter.java [hunk 4]
```diff
@@ -1434,6 +1434,7 @@ protected void adjustSpawn() {
      */
     protected abstract void prepareGenerator();
 
+    // TODO Fix broken getConstructor call
```
Comment (geNAZt) [inline gomint-server/src/main/java/io/gomint/server/world/WorldAdapter.java]: Whats broken with it? Should it fixed in this PR? Is this still needed with the new registry?

**Review by geNAZt — CHANGES_REQUESTED** [review]

File: gomint-server/src/main/java/io/gomint/server/world/WorldAdapter.java (hunk 4 above)
Comment (ghost) [reply gomint-server/src/main/java/io/gomint/server/world/WorldAdapter.java]: Nah, its not broken. Forgot to remove that.

File: gomint-api/src/main/java/io/gomint/world/generator/ChunkGeneratorRegistry.java (hunk 3 above)
Comment (ghost) [reply gomint-api/src/main/java/io/gomint/world/generator/ChunkGeneratorRegistry.java]: Optional is always better than returning a null reference instead but for the sake of concistency I'm going to remove that.

File: gomint-api/src/main/java/io/gomint/world/generator/ChunkGeneratorRegistry.java (hunk 2 above)
Comment (ghost) [reply gomint-api/src/main/java/io/gomint/world/generator/ChunkGeneratorRegistry.java]: Good point.

File: gomint-api/src/main/java/io/gomint/world/generator/ChunkGenerator.java (hunk 1 above)
Comment (ghost) [reply gomint-api/src/main/java/io/gomint/world/generator/ChunkGenerator.java]: Changed in no time

**Comment by ghost** [comment]
There you go. Your changes you requested.

**Pull request merged**